language/move-vm/runtime/src/interpreter.rs

```diff
@@ -869,6 +1033,504 @@ impl Frame {
+                interpreter.operand_stack.push_ty(Type::U16)?;
+            }
+            Bytecode::CastU32 => {
+                interpreter.operand_stack.pop_ty()?;
+                interpreter.operand_stack.push_ty(Type::U32)?;
+            }
+            Bytecode::CastU64 => {
+                interpreter.operand_stack.pop_ty()?;
+                interpreter.operand_stack.push_ty(Type::U64)?;
+            }
+            Bytecode::CastU128 => {
+                interpreter.operand_stack.pop_ty()?;
+                interpreter.operand_stack.push_ty(Type::U128)?;
+            }
+            Bytecode::CastU256 => {
+                interpreter.operand_stack.pop_ty()?;
+                interpreter.operand_stack.push_ty(Type::U256)?;
+            }
+            Bytecode::Add
+            | Bytecode::Sub
+            | Bytecode::Mul
+            | Bytecode::Mod
+            | Bytecode::Div
+            | Bytecode::BitOr
+            | Bytecode::BitAnd
+            | Bytecode::Xor
+            | Bytecode::Or
+            | Bytecode::And => {
+                let lhs = interpreter.operand_stack.pop_ty()?;
+                let rhs = interpreter.operand_stack.pop_ty()?;
+                lhs.check_eq(&rhs)?;
+                interpreter.operand_stack.push_ty(lhs)?;
+            }
+            Bytecode::Shl | Bytecode::Shr => {
+                interpreter.operand_stack.pop_ty()?;
+                let rhs = interpreter.operand_stack.pop_ty()?;
+                interpreter.operand_stack.push_ty(rhs)?;
+            }
+            Bytecode::Lt | Bytecode::Le | Bytecode::Gt | Bytecode::Ge => {
+                let lhs = interpreter.operand_stack.pop_ty()?;
+                let rhs = interpreter.operand_stack.pop_ty()?;
+                lhs.check_eq(&rhs)?;
+                interpreter.operand_stack.push_ty(Type::Bool)?;
+            }
+            Bytecode::Eq | Bytecode::Neq => {
+                let lhs = interpreter.operand_stack.pop_ty()?;
+                let rhs = interpreter.operand_stack.pop_ty()?;
+                if lhs != rhs {
+                    return Err(
+                        PartialVMError::new(StatusCode::UNKNOWN_INVARIANT_VIOLATION_ERROR)
+                            .with_message(
+                                "Integer binary operation expecting values of same type"
+                                    .to_string(),
+                            ),
+                    );
+                }
+                check_ability(resolver.loader().abilities(&lhs)?.has_drop())?;
+                interpreter.operand_stack.push_ty(Type::Bool)?;
+            }
+            Bytecode::MutBorrowGlobal(idx) => {
+                interpreter
+                    .operand_stack
+                    .pop_ty()?
+                    .check_eq(&Type::Address)?;
+                let ty = resolver.get_struct_type(*idx);
+                interpreter
+                    .operand_stack
+                    .push_ty(Type::MutableReference(Box::new(ty)))?;
```

Comment:
I think we want to check for the `key` ability in all of the global storage operators. The bytecode verifier does this, anway.